akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorStashSpec.scala

```diff
@@ -0,0 +1,443 @@
+      value2 should ===(5200)
+      value3 should ===(8100)
+    }
+
+    // FIXME what is desired behavior?
+    "discard stashing state when restarted" in {
+      val c = spawn(counter(nextPid()))
+      val ackProbe = TestProbe[Ack]
+      val stateProbe = TestProbe[State]
+
+      c ! Increment("inc-1", ackProbe.ref)
+      ackProbe.expectMessage(Ack("inc-1"))
+
+      c ! Deactivate("deact", ackProbe.ref)
+      ackProbe.expectMessage(Ack("deact"))
+
+      c ! Increment("inc-2", ackProbe.ref)
+      c ! Increment("inc-3", ackProbe.ref)
+      c ! GetValue(stateProbe.ref)
+      stateProbe.expectMessage(State(1, Vector(0), active = false))
+
+      c ! Throw("throw", new TestException("test"), ackProbe.ref)
+      ackProbe.expectMessage(Ack("throw"))
+
+      c ! Increment("inc-4", ackProbe.ref)
+      c ! GetValue(stateProbe.ref)
+      stateProbe.expectMessage(State(1, Vector(0), active = false))
+
+      c ! Activate("act", ackProbe.ref)
+      c ! Increment("inc-5", ackProbe.ref)
+
+      ackProbe.expectMessage(Ack("act"))
+      // inc-2 an inc-3 was in external stash, and didn't survive restart, as expected
+      ackProbe.expectMessage(Ack("inc-4"))
+      ackProbe.expectMessage(Ack("inc-5"))
+
+      c ! GetValue(stateProbe.ref)
+      stateProbe.expectMessage(State(3, Vector(0, 1, 2), active = true))
+    }
+
+    // FIXME what is desired behavior?
+    "keep internal stashing state when persist failed" in {
+      val c = spawn(counter(PersistenceId("fail-fifth-a")))
+      val ackProbe = TestProbe[Ack]
+      val stateProbe = TestProbe[State]
+
+      (1 to 10).foreach { n ⇒
+        c ! Increment(s"inc-$n", ackProbe.ref)
+      }
+
+      (1 to 10).foreach { n ⇒
+        if (n != 5)
+          ackProbe.expectMessage(Ack(s"inc-$n"))
+      }
+
+      c ! GetValue(stateProbe.ref)
+      stateProbe.expectMessage(State(9, Vector(0, 1, 2, 3, 4, 5, 6, 7, 8), active = true))
+    }
+
+    // FIXME what is desired behavior?
+    "discard external stashing state when persist failed" in {
+      val c = spawn(counter(PersistenceId("fail-fifth-b")))
+      val ackProbe = TestProbe[Ack]
+      val stateProbe = TestProbe[State]
+
+      c ! Increment("inc-1", ackProbe.ref)
+      ackProbe.expectMessage(Ack("inc-1"))
+
+      c ! Deactivate("deact", ackProbe.ref)
+      ackProbe.expectMessage(Ack("deact"))
+
+      c ! Increment("inc-2", ackProbe.ref)
+      c ! Increment("inc-3", ackProbe.ref) // this will fail when unstashed
+      c ! Increment("inc-4", ackProbe.ref)
+      c ! Increment("inc-5", ackProbe.ref)
+      c ! GetValue(stateProbe.ref)
+      stateProbe.expectMessage(State(1, Vector(0), active = false))
+
+      c ! Activate("act", ackProbe.ref)
+      c ! Increment("inc-6", ackProbe.ref)
+
+      ackProbe.expectMessage(Ack("act"))
+      ackProbe.expectMessage(Ack("inc-2"))
+      // inc-3 failed
+      // inc-4, inc-5 still in external stash buffer but UnstashAll in progress was aborted -- not desired
```

Comment:
I think we could rather easily preserve the complete external stash state for persist failures, if that is what we want